bin/rpc/rpc_common.ml

```diff
@@ -36,6 +36,51 @@ let client_term builder f =
   Scheduler.go_with_rpc_server ~common ~config f
 ;;
 
+let retry_loop once =
+  let open Fiber.O in
+  let rec loop () =
+    let* res = once () in
+    match res with
+    | Some result -> Fiber.return result
+    | None ->
+      let* () = Scheduler.sleep ~seconds:0.2 in
+      loop ()
+  in
+  loop ()
+;;
+
+let establish_connection_or_raise ~wait once =
+  let open Fiber.O in
+  if wait
+  then retry_loop once
+  else
+    let+ res = once () in
+    match res with
+    | Some conn -> conn
+    | None ->
+      let (_ : Dune_rpc_private.Where.t) = active_server () in
+      User_error.raise
+        [ Pp.text "failed to establish connection even though server seems to be running"
+        ]
+;;
```

Comment:
I think this would be kind of better if there were `establish_connection_with_retry` and `establish_connection_exn`. This way it is these two functions joined with an `if`. It would be more readable to dispatch in the caller.